rsc/client/valve/replay.go

```diff
@@ -0,0 +1,92 @@
+package valve
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+)
+
+var (
+	ErrDownloadRelpay = fmt.Errorf("failed to download the dota2 replay")
+)
+
+type ReplayDownloader interface {
+	// Download
+	// downloading the replay file is downloaded directly to the
+	// Valve replay proxy using the replayURL.
+	//
+	// replayURL contains the full raw URI to download the replay to the Valve replay proxy,
+	// for example http://replay153.valve.net/570/7132230434_1635105612.dem.bz2
+	// Based on that value, we can download the replay DotA2 file directly using that URL.
+	//
+	// destination parameter specifies the file destination to save the downloaded replay file.
+	Download(ctx context.Context, replayURL string, destination string) error
+}
+
+type replay struct {
+	httpClient *http.Client
+}
+
+func NewReplay(httpClient *http.Client) ReplayDownloader {
```

Comment:
not a fan of exposing interface but hiding implementation, prefer the other way around. and why do we need an interface here in the first place?

Reply:
> not a fan of exposing interface but hiding implementation, prefer the other way around
I agree with this for a specific reason. However, when we need to reuse the abstraction, it is useful to use an `interface`.

> and why do we need an interface here in the first place?
Mmm, I'm just thinking of creating an abstraction for the `Replay Downloader` and reusing the interface later when we need to download other implementations.

So, do you prefer to expose directly to `struct` instead of using `interface`?

Reply:
just return struct and remove the interface, unless we plan to have multiple implementstions of `ReplayDownloader`, do we? for other cases where we have multiple implementations of an interface, this is fine though, as suggested by [the first paragraph here](https://go.dev/doc/effective_go#generality).

Reply:
I'll go to return directly to struct instead of interface

Reply:
updated